runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/smithy/Document.kt

```diff
@@ -4,161 +4,185 @@
+                        null -> Null
+                        else -> throw IllegalArgumentException("incompatible value $it used to construct document list")
+                    }
+                }
+            )
+    }
 
     /**
-     * Returns the content as a byte which may involve rounding
+     * Wraps a [kotlin.Number].
      */
-    val byte: Byte get() = content.toByte()
+    data class Number(val value: kotlin.Number) : Document() {
+        override fun toString() = value.toString()
+    }
 
     /**
-     * Returns the content as a int which may involve rounding
+     * Wraps a [kotlin.String].
      */
-    val int: Int get() = content.toInt()
+    data class String(val value: kotlin.String) : Document() {
+        override fun toString() = "\"$value\""
+    }
 
     /**
-     * Returns the content as a long which may involve rounding
+     * Wraps a [kotlin.Boolean].
      */
-    val long: Long get() = content.toLong()
+    data class Boolean(val value: kotlin.Boolean) : Document() {
+        override fun toString() = value.toString()
+    }
 
     /**
-     * Returns the content as a float which may involve rounding
+     * Wraps a [kotlin.collections.List].
      */
-    val float: Float get() = content.toFloat()
+    data class List(val value: kotlin.collections.List<Document>) : Document() {
+        override fun toString() = value.joinToString(separator = ",", prefix = "[", postfix = "]")
+    }
 
     /**
-     * Returns the content as a double which may involve rounding
+     * Wraps a [kotlin.collections.Map].
      */
-    val double: Double get() = content.toDouble()
-
-    override fun toString(): String = content.toString()
-}
+    data class Map(val value: kotlin.collections.Map<kotlin.String, Document>) : Document() {
+        override fun toString() = value
+            .entries
+            .joinToString(
+                separator = ",",
+                prefix = "{",
+                postfix = "}",
+                transform = { (k, v) -> """"$k":$v""" }
+            )
+    }
 
-/**
- * Class representing document `array` type
- */
-data class SmithyArray(val content: List<Document>) : Document(), List<Document> by content {
     /**
-     * Returns [index] th element of an array as [SmithyNumber] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
+     * Represents a `null` value.
      */
-    fun getNumber(index: Int) = content[index] as? SmithyNumber
+    object Null : Document() {
+        override fun toString() = "null"
+    }
 
-    /**
-     * Returns [index] th element of an array as [SmithyBool] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getBoolean(index: Int) = content[index] as? SmithyBool
+    fun asNumber() = (this as Number).value
```

Comment:
all of these should have explicit return types (which reminds me we really need to turn that check on...)